### DISCUSSION TITLE
Nested `with` expressions

### DISCUSSION THREAD

**Discussion by AFatNiBBa** [dotnet/csharplang#8319] [Language Ideas]
## Example
Let's assume we have these two `struct`s
```cs
public struct Example1
{
    public int Value1;
    public Example2 Inner;

    public Example1() => Inner = new() { Value4 = 4 };
}

public struct Example2
{
    public int Value2;
    public int Value3;
    public int Value4;
}
```
It is possible for me to initialize `Example1` like this
```cs
var a = new Example1
{
    Value1 = 1,
    Inner = { Value2 = 2, Value3 = 3 }
};
```
But if I write a `with` expression, I won't be able to MODIFY `Example1.Inner`
```cs
var b = a with
{
    Value1 = 4,
    Inner = { Value2 = 5, Value3 = 6 } // Unallowed!
};
```
I assume this is because if `Example2` were to be a `class`, that would have changed the original `Example2` instance stored in `a.Inner`

## Proposal
I think that it would be beneficial for nested sets of `{ }` inside a `with` expression to be converted in additional `with` expressions like so
```cs
var b = a with
{
    Value1 = 4,
    Inner = a.Inner with { Value2 = 5, Value3 = 6 }
};
```
This would allow the modification only for clonable objects (`record`, `struct`) and wouldn't work on standard reference types

**Comment by colejohnson66** [comment]
#8289
